### PR TITLE
added preprocessing in blobFromImage for SSD

### PR DESCRIPTION
**Pull request by haritha1313**
resolves https://github.com/opencv/opencv/issues/9967

### This pullrequest changes

changed comments related to blobfromimage() in ssd_object_detection sample of DNN. Implemented associated preprocessing steps within blobfromimage().

### CODE REVIEW

File: modules/dnn/include/opencv2/dnn/dnn.hpp [hunk 1]
```diff
@@ -712,7 +712,7 @@ CV__DNN_EXPERIMENTAL_NS_BEGIN
      *  @returns 4-dimansional Mat with NCHW dimensions order.
      */
     CV_EXPORTS_W Mat blobFromImage(const Mat& image, double scalefactor=1.0, const Size& size = Size(),
-                                   const Scalar& mean = Scalar(), bool swapRB=true, bool crop=true);
+                                   const Scalar& mean = Scalar(), bool swapRB=true, bool preprocess=false, bool crop=true);
```
Comment (dkurt) [inline modules/dnn/include/opencv2/dnn/dnn.hpp]: `blobFromImage` is already a kind of image preprocessing. So `preprocess=false` means do nothing I guess. I think it isn't necessary.

File: samples/dnn/ssd_object_detection.cpp [hunk 2]
```diff
@@ -126,9 +96,8 @@ int main(int argc, char** argv)
             cvtColor(frame, frame, COLOR_BGRA2BGR);
 
         //! [Prepare blob]
-        Mat preprocessedFrame = preprocess(frame);
-
-        Mat inputBlob = blobFromImage(preprocessedFrame, 1.0f, Size(), Scalar(), false); //Convert Mat to batch of images
+        //preprocess and convert frame to a series of images
+        Mat inputBlob = blobFromImage(frame, 1.0f, Size(), Scalar(), false, true);  //returns 4-dimensional blob
```
Comment (dkurt) [inline samples/dnn/ssd_object_detection.cpp]: At the https://github.com/opencv/opencv/issues/9967#issuecomment-343419551 I meant to use `blobFromImage` to do equal preprocessing steps (resize, subtract mean).

**Comment by dkurt** [comment]
@haritha1313, thank you for your contribution! There are some moments that need to be resolved. Please, read comments.

**Comment by haritha1313** [comment]
@dkurt  Thank you for the review. It was a slight misunderstanding which I will correct and get back soon.

File: samples/dnn/squeezenet_halide.cpp [hunk 3]
```diff
@@ -79,8 +79,7 @@ int main(int argc, char **argv)
         exit(-1);
     }
 
-    resize(img, img, Size(227, 227));                // SqueezeNet v1.1 predict class by 3x227x227 input image.
-    Mat inputBlob = blobFromImage(img, 1.0, Size(), Scalar(), false);  // Convert Mat to 4-dimensional batch.
+    Mat inputBlob = blobFromImage(img, 1.0, Size(227, 227), Scalar(), false);  // Convert Mat to 4-dimensional batch.
```
Comment (dkurt) [inline samples/dnn/squeezenet_halide.cpp]: Please consider the last flag `crop` of `blobFromImage`. By default it's true that means we keep aspect ratios of image by cropping some pixels from the highest dimension. So to get similar input we should pass `false` value. https://docs.opencv.org/master/d6/d0f/group__dnn.html#ga33d1b39b53a891e98a654fdeabba22eb

File: samples/dnn/ssd_object_detection.cpp [hunk 4]
```diff
@@ -126,9 +96,7 @@ int main(int argc, char** argv)
             cvtColor(frame, frame, COLOR_BGRA2BGR);
 
         //! [Prepare blob]
-        Mat preprocessedFrame = preprocess(frame);
-
-        Mat inputBlob = blobFromImage(preprocessedFrame, 1.0f, Size(), Scalar(), false); //Convert Mat to batch of images
+        Mat inputBlob = blobFromImage(frame, 1.0f, Size(300, 300), Scalar(104, 117, 123), false); //Convert Mat to batch of images
```
Comment (dkurt) [inline samples/dnn/ssd_object_detection.cpp]: The same: `crop = false`.

File: samples/dnn/tf_inception.cpp [hunk 5]
```diff
@@ -78,12 +78,7 @@ int main(int argc, char **argv)
         exit(-1);
     }
 
-    cv::Size inputImgSize = cv::Size(224, 224);
-
-    if (inputImgSize != img.size())
-        resize(img, img, inputImgSize);       //Resize image to input size
-
-    Mat inputBlob = blobFromImage(img);   //Convert Mat to image batch
+    Mat inputBlob = blobFromImage(img, 1.0f, Size(224, 224));   //Convert Mat to batch of images
```
Comment (dkurt) [inline samples/dnn/tf_inception.cpp]: `crop = false` so we have to add an empty `Scalar()` for means and `swapRB=true`

File: samples/dnn/yolo_object_detection.cpp [hunk 6]
```diff
@@ -104,13 +101,8 @@ int main(int argc, char** argv)
...
             cvtColor(frame, frame, COLOR_BGRA2BGR);
 
-        //! [Resizing without keeping aspect ratio]
-        Mat resized;
-        resize(frame, resized, Size(network_width, network_height));
-        //! [Resizing without keeping aspect ratio]
-
         //! [Prepare blob]
-        Mat inputBlob = blobFromImage(resized, 1 / 255.F); //Convert Mat to batch of images
+        Mat inputBlob = blobFromImage(frame, 1 / 255.F, Size(416, 416)); //Convert Mat to batch of images
```
Comment (dkurt) [inline samples/dnn/yolo_object_detection.cpp]: The same as above: `Scalar()`, `swapRB=true` and `crop=false`

File: doc/tutorials/dnn/dnn_googlenet/dnn_googlenet.markdown [hunk 7]
```diff
@@ -37,11 +37,9 @@ Explanation
 
 -# Read input image and convert to the blob, acceptable by GoogleNet
    @snippet dnn/caffe_googlenet.cpp Prepare blob
-   Firstly, we resize the image and change its channel sequence order.
+   Firstly, we resize the image and perform mean subtraction using cv::dnn::blobFromImage constructor
 
-   Now image is actually a 3-dimensional array with 224x224x3 shape.
-
-   Next, we convert the image to 4-dimensional blob (so-called batch) with 1x3x224x224 shape by using special cv::dnn::blobFromImages constructor.
+   The constructor then converts the image to a 4-dimensional blob (so-called batch) with 1x3x224x224 shape by using special cv::dnn::blobFromImages constructor.
```
Comment (dkurt) [inline doc/tutorials/dnn/dnn_googlenet/dnn_googlenet.markdown]: There is no `cv::dnn::blobFromImages` invocation here. So we can just say that we make 4-dimensional blob with size `1x3x224x224` applying all necessary pre-processing procedures like resizing and mean subtraction.

**Comment by dkurt** [comment]
@haritha1313, Finally, may I ask you to squash all changes into a single commit?

**Comment by haritha1313** [comment]
@dkurt  Sorry for the mess, im not used to some commands like rebase.

**Comment by dkurt** [comment]
:+1: looks good for me. Thank you!

**Pull request merged**